cirq/study/resolver.py

```diff
@@ -121,33 +130,57 @@ def value_of(self, value: Union['cirq.TParamKey', float]) -> 'cirq.TParamVal':
         # If sympy did its job and wasn't slower than molasses,
         # we wouldn't need the following block.
         if isinstance(value, sympy.Add):
-            summation = self.value_of(value.args[0])
+            summation = self.value_of(value.args[0], recursive)
             for addend in value.args[1:]:
-                summation += self.value_of(addend)
+                summation += self.value_of(addend, recursive)
             return summation
         if isinstance(value, sympy.Mul):
-            product = self.value_of(value.args[0])
+            product = self.value_of(value.args[0], recursive)
             for factor in value.args[1:]:
-                product *= self.value_of(factor)
+                product *= self.value_of(factor, recursive)
             return product
         if isinstance(value, sympy.Pow) and len(value.args) == 2:
-            return np.power(self.value_of(value.args[0]), self.value_of(value.args[1]))
+            return np.power(
+                self.value_of(value.args[0], recursive), self.value_of(value.args[1], recursive)
+            )
+
+        if not isinstance(value, sympy.Basic):
+            # No known way to resolve this variable, return unchanged.
+            return value
 
         # Input is either a sympy formula or the dictionary maps to a
         # formula.  Use sympy to resolve the value.
         # Note that sympy.subs() is slow, so we want to avoid this and
         # only use it for cases that require complicated resolution.
-        if isinstance(value, sympy.Basic):
-            v = value.subs(self.param_dict)
+        if not recursive:
+            # Resolves one step at a time. For example:
+            # a.subs({a: b, b: c}) == b
+            v = value.subs(self.param_dict, simultaneous=True)
             if v.free_symbols:
                 return v
             elif sympy.im(v):
                 return complex(v)
             else:
                 return float(v)
 
-        # No known way to resolve this variable, return unchanged.
-        return value
+        # Recursive parameter resolution. We can safely assume that value is a
+        # single symbol, since combinations are handled earlier in the method.
+        try:
+            v = self._deep_eval_map[value]
+            if v is not None:
+                return v
+            raise RecursionError('Evaluation of {value} indirectly contains itself.')
+        except KeyError:
```

Comment:
This isn't really an error, but one of the normal branches of the algorithm. We should handle it with a regular conditional:

```python
if value in self._deep_eval_map:
    v = self._deep_eval_map[value]
    if v is not None:
        return v
    raise RecursionError(...)
self._deep_eval_map[value] = None
```

Reply:
I had the impression that this usage of `try...except` was the Pythonic way of doing "if x in dict, get x". Open to changing it if there's an issue that I'm missing here.

Reply:
I disagree that triggering a `KeyError` is a pythonic way to check whether a key exists in a dict. Python provides the `in` operator to do this directly and cleanly. BTW: A second pythonic way is the `dict.get(key, default)` function, but I don't think it's very helpful here.

A try-except block is suitable to handle actual errors, not normal control flow. In particular, it should be used when errors may originate at various function call depth and so local checks cannot detect error condition. Also, handling errors this way comes with costs such as the need to instantiate the exception, to do an `isinstance` check and to potentially unwind the stack. There is also the readability cost due to the clutter introduced by extra keywords and type names.

Reply:
BTW: I checked whether we do this in cirq and only found four instances. I didn't count the number of times we use the `in` operator directly, but searched for `'if .* in [a-zA-Z_]*:'` as a rough proxy. Found 87 instances. Thus, even by the standard of consistency with cirq coding style, we should not be triggering `KeyError` here.

Reply:
Fair enough. I wasn't able to find a source for my impression either, so I've made the change.

Reply:
As @mpharrigan points out [here](https://github.com/quantumlib/Cirq/pull/3568#issuecomment-739070461), I was wrong about try-except being a non-pythonic way of doing this.

Except for the four cases which I changed in #3568, we had generally been avoiding using try-except blocks to check key presence, so I still think we should prefer the `in` operator for local consistency and also for other reasons listed above (efficiency and readability costs and, in some instances, the risk of catching exceptions raised deeper down the call stack). That said, I'm open to changing my mind if people want to do it the EAFP way going forward.

Reply:
I don't actually like EAFP, so works for me